File: packages/test-plugin/android/src/main/java/io/capkit/test/TestConfig.kt
package io.capkit.test

import android.content.Context
import com.getcapacitor.Plugin

/**
 * Plugin configuration holder.
 *
 * This class is responsible for reading and parsing configuration values
 * defined under the `Test` key in `capacitor.config.ts`.
 *
 * Configuration is read once during plugin initialization and treated as
 * immutable runtime input.
 */
class TestConfig(plugin: Plugin) {
  /**
   * Android application context.
   * Exposed for native components that may require it.
   */
  val context: Context

  /**
   * Enables verbose / debug logging for the plugin.
   *
   * When enabled, additional logs are printed to Logcat via [Logger.debug].
   *
   * Default: false
   */
  val verboseLogging: Boolean

  /**
   * Optional custom message appended to echoed values.
   *
   * This value is provided for demonstration purposes and shows how to
   * pass static configuration from JavaScript to native code.
   */
  val customMessage: String

  init {
    context = plugin.context

    val config = plugin.getConfig()

    // Parse verboseLogging (default: false)
    verboseLogging = config.getBoolean("verboseLogging", false)

    // Parse customMessage (default fallback applied)
    val cm = config.getString("customMessage")
    customMessage = if (cm.isNullOrBlank()) " (from config)" else cm
  }
}

File: packages/test-plugin/ios/Sources/TestPlugin/TestConfig.swift
import Foundation
import Capacitor

/**
 Plugin configuration container.

 This struct is responsible for reading and exposing static configuration
 values defined under the `Test` key in `capacitor.config.ts`.

 Configuration is:
 - read once during plugin initialization
 - treated as immutable runtime input
 - consumed only by native code (never by JavaScript)
 */
public struct TestConfig {

    // MARK: - Configuration Keys

    // Define configuration keys for consistency
    private struct Keys {
        static let verboseLogging = "verboseLogging"
        static let customMessage = "customMessage"
    }

    // MARK: - Public Config Values

    /**
     Enables verbose native logging.

     When enabled, additional debug information is printed
     to the Xcode console via the plugin logger.

     Default: false
     */
    public let verboseLogging: Bool

    /**
     Custom message appended to echoed values.

     This property exists primarily for demonstration purposes and
     shows how to pass static configuration values from JavaScript
     to native code.

     Default: " (from config)"
     */
    public let customMessage: String?

    // MARK: - Private Defaults

    // Default values
    private let defaultVerboseLogging = false
    private let defaultCustomMessage = " (from config)"

    // MARK: - Init

    /**
     NOTE:
     We intentionally use `getConfigValue(_:)` instead of `getConfig()`
     due to Swift Package Manager toolchain limitations observed in
     Capacitor v8 (PluginConfig not reliably accessible in some setups).

     Initializes the configuration by reading values from the Capacitor bridge.

     - Parameter plugin: The CAPPlugin instance used to access typed configuration.
     */
    init(plugin: CAPPlugin) {
        let pluginId = plugin.pluginId

        // Bool
        verboseLogging = plugin.getConfigValue(Keys.verboseLogging) as? Bool ?? defaultVerboseLogging

        // String
        customMessage = plugin.getConfigValue(Keys.customMessage) as? String ?? defaultCustomMessage
    }
}
